Tests/LeafKitTests/LeafKitTests.swift

```diff
@@ -917,23 +823,170 @@ final class LeafKitTests: XCTestCase {
+        test.files["/b.leaf"] = "#extend(\"c\")"
+        test.files["/c.leaf"] = "#extend(\"d\")"
+        
+        let renderer = LeafRenderer(
+            configuration: .init(rootDirectory: "/"),
+            cache: DefaultLeafCache(),
+            files: test,
+            eventLoop: EmbeddedEventLoop()
+        )
+        
+        do {
+            _ = try renderer.render(path: "a", context: [:]).wait()
+            XCTFail("Should have thrown LeafError.cyclicalReference")
+        } catch {
+            guard let e = error as? LeafError else { XCTFail("Wrong error: \(error.localizedDescription)"); return }
+            // list of chain references may be out of order
+            XCTAssertEqual(e.localizedDescription, LeafError.noTemplateExists("/d.leaf").localizedDescription)
+        }
+    }
+    
+    func testImportResolve() {
+        var test = TestFiles()
+        test.files["/a.leaf"] = """
+        #extend("b"):
+        #export("variable"):Hello#endexport
+        #endextend
+        """
+        test.files["/b.leaf"] = """
+        #import("variable")
+        """
+        
+        let renderer = LeafRenderer(
+            configuration: .init(rootDirectory: "/"),
+            cache: DefaultLeafCache(),
+            files: test,
+            eventLoop: EmbeddedEventLoop()
+        )
+        
+        do {
+            let output = try renderer.render(path: "a", context: [:]).wait().string
+            XCTAssertEqual(output, "Hello")
+        } catch {
+            let e = error as! LeafError
+            XCTFail(e.localizedDescription)
+        }
+    }
+    
+    
+    func testCacheSpeedLinear() {
+        self.measure {
+            self._testCacheSpeedLinear(templates: 10, iterations: 100)
+        }
+    }
+    
+    func _testCacheSpeedLinear(templates: Int, iterations: Int) {
+        let group = MultiThreadedEventLoopGroup(numberOfThreads: System.coreCount)
+        var test = TestFiles()
+       
+        for name in 1...templates { test.files["/\(name).leaf"] = "Template /\(name).leaf" }
+        let renderer = LeafRenderer(
+            configuration: .init(rootDirectory: "/"),
+            cache: DefaultLeafCache(),
+            files: test,
+            eventLoop: group.next()
+        )
+        
+        for iteration in 1...iterations {
+            let key = String((iteration % templates) + 1)
+            _ = renderer.render(path: key, context: [:])
+        }
+        
+        group.shutdownGracefully { shutdown in
+            guard shutdown == nil else { XCTFail("ELG shutdown issue"); return }
+            XCTAssertEqual(renderer.cache.entryCount(), templates)
+        }
+    }
+    
+    func testCacheSpeedRandom() {
+        self.measure {
+            // layer1 > layer2 > layer3
+            self._testCacheSpeedRandom(layer1: 100, layer2: 20, layer3: 10, iterations: 130)
+        }
+    }
+    
+    func _testCacheSpeedRandom(layer1: Int, layer2: Int, layer3: Int, iterations: Int) {
```

Comment:
This is excellent, I love it!

Reply:
RE these speeds tests and others, I have various updates to unify testing as well but don't want to try and merge those until the big stuff is worked out so it's a little clearer